activesupport/lib/active_support/core_ext/class/attribute.rb

```diff
@@ -92,50 +92,63 @@ def class_attribute(
     instance_predicate: true,
     default: nil
   )
+    class_method_def = []
+    method_def = []
+
     attrs.each do |name|
-      singleton_class.silence_redefinition_of_method(name)
-      define_singleton_method(name) { default }
+      unless name.is_a?(Symbol) || name.is_a?(String)
+        raise TypeError, "#{name.inspect} is not a symbol nor a string"
+      end
 
-      singleton_class.silence_redefinition_of_method("#{name}?")
-      define_singleton_method("#{name}?") { !!public_send(name) } if instance_predicate
+      class_method_def <<
+        "silence_redefinition_of_method(:#{name})" <<
+        "def #{name}" << # In case the method exists and is not public
+        "end"
 
-      ivar = "@#{name}".to_sym
+      class_method_def <<
+        "silence_redefinition_of_method(:#{name}=)" <<
+        "def #{name}=(value)" <<
+        "  redefine_method(:#{name}) { value } if singleton_class?" <<
+        "  redefine_singleton_method(:#{name}) { value }" <<
+        "  value" <<
+        "end"
 
-      singleton_class.silence_redefinition_of_method("#{name}=")
-      define_singleton_method("#{name}=") do |val|
-        redefine_singleton_method(name) { val }
+      if instance_predicate
+        class_method_def <<
+          "silence_redefinition_of_method(:#{name}?)" <<
+          "def #{name}?" <<
+          "  !!self.#{name}" <<
+          "end"
 
-        if singleton_class?
-          class_eval do
-            redefine_method(name) do
-              if instance_variable_defined? ivar
-                instance_variable_get ivar
-              else
-                singleton_class.send name
-              end
-            end
-          end
-        end
-        val
+        method_def <<
```

Comment:
I think placing the instance predicate method here changes the semantics of this method.

In the code before, the instance predicate was only defined if `instance_reader` was truthy (as well as `instance_predicate`).

In this PR, as long as `instance_predicate` is defined, the instance predicate method is defined, which seems to be a behaviour change.

Reply:
Nice catch. I fixed it.